content/languages/python/authoring/index.md

```diff
@@ -0,0 +1,252 @@
+
+### Reference solution
+
+If the test suite happens to use a reference solution to calculate expected values (which [should be avoided](/authoring/guidelines/submission-tests/#reference-solution), but is not always possible), or some kind of reference data, precalculated arrays, etc., it can be redefined, overwritten, or accessed directly by the user solution. To avoid this, it should be defined in a scope local to the testing function, `it` block, or `describe` block. The reference solution or data _must not_ be defined in the top-level scope of the test suite, nor in the [Preloaded code](/authoring/guidelines/preloaded/).
+
+### Calling assertions
+
+The Python testing framework provides a set of useful [assertions](/languages/python/codewars-test/#assertions), but when used incorrectly, they can cause a series of problems:
+- Stacktraces of a crashing user solution can reveal details that should not be visible,
+- Use of an assertion not suitable for the given case may lead to incorrect test results,
+- Incorrectly used assertions may produce confusing or unhelpful messages.
+
+To avoid the above problems, calls to assertion functions should respect the following rules:
+- The expected value should be calculated _before_ invoking an assertion. The `expected` parameter passed to the assertion should not be a function call expression, but a value calculated directly beforehand.
+- Appropriate assertion functions should be used for each given test. `assert_equals` is not suitable in all situations. Use `assert_approx_equals` for floating point comparisons, `expect` for tests on boolean values, `expect_error` to test error handling.
+- Some additional attention should be paid to the order of parameters passed to assertion functions. It differs between various assertion libraries, and it happens to be quite often confused by authors, mixing up `actual` and `expected` in assertion messages. For the Python testing framework, the order is `(actual, expected)`.
+- One somewhat distinctive feature of Python assertions is that by default, a failed assertion does not cause a test case to fail early. It can lead to unexpected crashes when an actual value had already been asserted to be invalid, but the execution of the current test case was not stopped and following assertions continue to refer to it. This behavior can be overridden by passing the `allow_raise=True` argument to the assertion functions that support it.
+- To avoid unexpected crashes in tests, it's recommended to perform some additional assertions before assuming that the answer returned by the user solution has some particular type, form, or value. For example, if the test suite sorts the returned list to verify its correctness, an explicit assertion should be added to check whether the returned object is actually a list, and not, for example, `None`.
+
+
+## Additional restrictions
+
+### Accessing the solution file
```

Comment:
there is always that problem about when to do the length check... In some languages you have to do the test in the test cases part to not allow the `JS/return;` cheat, but in python, that's not a problem, afaik. => should we add that the length check must be done in the preloaded section? (so import of the test framewrork there too)

Reply:
This part will be moved to more general docs: one on accessing user solution file, and another on how additional restrictions are bad in general. Here I will add something on module forbidder, or totally skip it, IDK yet.

Reply:
the module forbidder is python specific, so if it must be talked about, that's here, I guess. I'm still reluctant at looking into that thing, since I've no idea how to resolve the last way Unnamed broke it...

Reply:
I removed this section. Module forbidded will be described in separate doc, and accessing solution file will be not python-specific.